src/object.rs

```diff
@@ -50,6 +50,28 @@ pub trait Cast: IsA<Object> {
         unsafe { T::from(self.into()) }
     }
 
+    /// Upcasts an object to a reference of its superclass or interface `T`.
+    ///
+    /// *NOTE*: This statically checks at compile-time if casting is possible. It is not always
+    /// known at compile-time, whether a specific object implements an interface or not, in which case
+    /// `upcast` would fail to compile. `dynamic_cast` can be used in these circumstances, which
+    /// is checking the types at runtime.
+    ///
+    /// # Example
+    ///
+    /// ```ignore
+    /// let button = gtk::Button::new();
+    /// let widget = button.upcast_ref::<gtk::Widget>();
+    /// ```
+    #[inline]
+    fn upcast_ref<T>(&self) -> &T
+    where T: StaticType + UnsafeFrom<ObjectRef> + Wrapper,
+          Self: IsA<T> {
+        unsafe {
+            mem::transmute(self)
```

Comment:
The object casts are safe because all the types have the same memory representation apart from their name and the phantom data type, and `IsA` is an unsafe trait so must only be implemented if it's actually a proper GObject wrapper type